### ISSUE TITLE
Pipeline updates fail when resource requests are set

### ISSUE THREAD

**Issue by itssimon** [pachyderm/pachyderm#4339]
**What happened?**:

I have 5 pipelines running sequentially, all doing CPU intensive work. I have set `parallelism` to 8 and I have two Kubernetes nodes with 8 cores each running. Each pipeline pod requests 1 CPU, which should be fine given that I have 16 cores available.

When I update the pipelines Pachyderm tries to temporarily launch all 40 pods, which fails due to the resource requests. So I end up with >20 pods in Pending status and everything comes to a standstill. This is even though there are no datums to process.

However, when I remove the resource requests, Kubernetes sometimes decides to schedule everything on one node making my pipelines super slow because they are competing for CPU resources.

**What you expected to happen?**:

Pachyderm launching all 40 pods for pipeline updates seems unnecessary unless there actually are datums to process, in which case they should be launched sequentially, not all at once.

**Environment?**:
- Kubernetes version: 1.14.6
- Pachyderm CLI and pachd server version: 1.9.8
- Cloud provider: AWS EKS

**Comment by JoeyZwicker** [comment]
Just to double check, presumably you have standby set for these pipelines, which is why they're spinning up and down right?

I think its a known issue that we aren't being as smart as we could be with standby and when things get spun up and down. It's not a hard thing to optimize though.

**Comment by itssimon** [comment]
Yes, that's exactly right. All pipelines have `standby: true` set. If it's not too hard to optimize it sounds like a good quick win for the next release! 😃

**Comment by jdoliner** [comment]
This is unfortunately quite hard to fix. The reason for this behavior is that when you update a pipeline it creates a new job to run the new version of the pipeline which requires spinning up the pods. In your case there are no datums to process so there isn't actually any work to do, but that's not generally true. A lot of the time a pipeline is updated and there are `failed` or `recovered` datums that need to be processed with this new version. It's also not really feasible to check if there are datums that need to be processed before we spin up the workers, because checking if they've all been processed is itself an expensive operation if there are a large number of datums so doing it in pachd runs the risk of exhausting pachd's resources and breaking everything. Creating worker pods to check if there's datums to be processed puts more stress on kubernetes and the underlying infrastructure to schedule the pods. But it ensures that those things stay isolated and can't bring down the cluster all together.

I think the most realistic thing we can do to make this not fail is making it so that multiple pipelines can be updated atomically. That should prevent the thundering herd where each pipeline spins up at the same time and instead make them spin up sequentially. I'm not sure if we'll be able to make it so they don't spin up at all though.

**Comment by JoeyZwicker** [comment]
@itssimon I think I misunderstood what you were originally asking. 

Are you updating all 5 pipelines? If so, then yes, we'd spin up all 40 pods. 

Or are you doing something like updating one of the pipeline at the top of the DAG and all of the downstream ones are spinning up too? Is that what you mean by "should be launched sequentially"?

**Comment by itssimon** [comment]
Yes, I'm updating all 5 pipelines. They all use the same Docker image, so when that changes I'm updating all pipelines accordingly.

> It's also not really feasible to check if there are datums that need to be processed before we spin up the workers, because checking if they've all been processed is itself an expensive operation if there are a large number of datums so doing it in pachd runs the risk of exhausting pachd's resources and breaking everything.

I don't understand why this is expensive when I can see it with `pachctl list job` within a few milliseconds? Why can't `pachd` use the same information to decide whether pods need to be spun up?

Anyway, do you have a recommended workaround for this?

**Comment by JoeyZwicker** [comment]
@itssimon When you do `list job`, you're seeing the total number of datums pretty quickly, but the process of going through and deciding which ones need to be `skipped` or `processed` is pretty expensive and happens while the job is running.

It might be doable (JD can comment) to see if the total datum count is 0 and not spin things up, but if there are non-zero number of datums, Pachyderm doesn't specifically know if there is new data or not before the job starts up the workers.

**Comment by JoeyZwicker** [comment]
Dealing with a resource constrained setup like this is always likely to have some trickiness to it. @jdoliner might have a better suggestion, but one option would be:

1. Pause all 5 pipelines (aka: `stop pipeline`). 
2. Update all 5 pipelines
3. Unpause them one by one as they each finish their work and go back into standby (aka: `start pipeline`)

In order for the above to work, we'll need to fix a few minor bugs first, #4294 and #4328. #4296 might get fixed along the same time which will help make this a bit less of a sharp edge too.

**Comment by jdoliner** [comment]
> I don't understand why this is expensive when I can see it with pachctl list job within a few milliseconds? Why can't pachd use the same information to decide whether pods need to be spun up?

This tells you that the previous job processed all of its datums, but to know that there's nothing to process you need to also know that the datums in this job are identical to those in the previous job. There's a few reasons they might not be identical:

- The new version of the pipeline might change its inputs, adding new datums.
- There might have been commits to the inputs since the last job ran which weren't processed because the pipeline was in a stopped state.
- There might have been failed or recovered datums, these can in theory be detected through the progress from the last job, but those progress values are designed to just surface information to users, they're not considered operationally critical, and if there's a failure in updating them we eventually move on and let the job succeed.

And there's probably a bunch of other ways for the datums to not be identical that I'm not thinking of. It's possible for us to check all of these conditions before spinning up the workers, but it's liable to be error prone, and the result of getting it wrong is that the pipeline doesn't spin up at all and no results get processed. So for now it makes sense for us to do the conservative thing and go through the same processing loop each time and check the datums. In the future we may be able to optimize this more, but we'll have to do so carefully for the reasons above.


@JoeyZwicker workaround is needlessly complicated, all you need to do is space out your updates a little bit, rather than updating all of the pipelines at once, just update them one at a time. No need to worry about stopping the pipelines. That should also mean that none of those issues impact the workaround.
